### PR TITLE
implementing multicurl node, adding graph optimizer that can restruct…

### PR DESCRIPTION
**Pull request by rjpolackwich**
…ure our current graphs based on load_url to instead return chunk indices and url in prep for multicurl mapping. NOT ready to merge yet

### CODE REVIEW

File: gbdxtools/ipe/interface.py [hunk 1]
```diff
@@ -43,6 +52,75 @@
...
+    if not keys:
+        keys = dsk.keys()
+    for k in keys:
+        if dsk[k][0] is load_url:
+            dsk2[k] = (dsk[k][1], (k[1:]))
+        else:
+            dsk2[k] = dsk[k]
+    return dsk2
+
+def load_urls(collection, token, shape=(8,256,256), timeout=0.1):
```
Comment (pramukta) [inline gbdxtools/ipe/interface.py]: This is good, but complex C like code and I think the complexity stems from incrementally processing the completed requests.  While this is nice behavior, we don't really need to do it here.

File: gbdxtools/ipe/interface.py [hunk 2]
```diff
@@ -43,6 +52,75 @@
...
+        _curl.index = index
+
+        handles.append(_curl)
+        mc.add_handle(_curl)
+
+    while nhandles:
+        ret = mc.select(timeout)
+        if ret == -1:
+            continue
+        while True:
```
Comment (pramukta) [inline gbdxtools/ipe/interface.py]: This is the main loop we want to run at the top level (but below any retry logic)

File: gbdxtools/ipe/interface.py [hunk 3]
```diff
@@ -43,6 +52,75 @@
...
+        _curl.fp = NamedTemporaryFile(prefix='gbdxtools', suffix=ext, delete=False)
+        _curl.setopt(pycurl.WRITEDATA, _curl.fp.file)
+        _curl.url = url
+        _curl.index = index
+
+        handles.append(_curl)
+        mc.add_handle(_curl)
+
+    while nhandles:
+        ret = mc.select(timeout)
```
Comment (pramukta) [inline gbdxtools/ipe/interface.py]: since we are waiting on all fds and we shouldn't need to select in a loop

File: gbdxtools/ipe/interface.py [hunk 4]
```diff
@@ -43,6 +52,75 @@
...
+    handles = []
+    for url, index in collection:
+        _, ext = os.path.splitext(urlparse(url).path)
+        _curl = pycurl.Curl()
+        _curl.setopt(pycurl.NOSIGNAL, 1)
+        _curl.setopt(pycurl.HTTPHEADER, ['Authorization: Bearer {}'.format(token)])
+        _curl.setopt(pycurl.CONNECTTIMEOUT, 30)
+        _curl.setopt(pycurl.TIMEOUT, 300)
+        _curl.setopt(pycurl.URL, url)
+        _curl.fp = NamedTemporaryFile(prefix='gbdxtools', suffix=ext, delete=False)
```
Comment (pramukta) [inline gbdxtools/ipe/interface.py]: are you sure about this one?  I could be wrong but it looks like you are overriding curls internal file pointer with a file object.  I don't think fp needs to be set, the file ref just needs to be maintained.

File: gbdxtools/ipe/interface.py [hunk 5]
```diff
@@ -43,6 +52,75 @@
...
+
+    while nhandles:
+        ret = mc.select(timeout)
+        if ret == -1:
+            continue
+        while True:
+            ret, nhandles = mc.perform()
+            if ret != pycurl.E_CALL_MULTI_PERFORM:
+                break
+        nqueued, successful, failed = mc.info_read()
```
Comment (pramukta) [inline gbdxtools/ipe/interface.py]: I think you can move this inside the above conditional, and key the while loop on nqueued > 0 or successful + failed < nhandles.

File: gbdxtools/ipe/interface.py [hunk 6]
```diff
@@ -43,6 +52,75 @@
...
+        for h in successful:
```
Comment (pramukta) [inline gbdxtools/ipe/interface.py]: we should be able to do this unconditionally (when we don't care about retry logic)

**Review by pramukta — COMMENTED** [review]
This is a first cut at review, only focused in on the multi node.  It's good stuff in general, though I think we should be able to get a bit more clarity for our case.  Let me know if you want to go through it.

File: gbdxtools/ipe/interface.py (hunk 1 above)
Comment (rjpolackwich) [reply gbdxtools/ipe/interface.py]: want to go over this block today?

Comment (pramukta) [reply gbdxtools/ipe/interface.py]: yeah sure lets do it

**Pull request merged**